libraries/AP_Mission/AP_Mission_Relative.cpp

```diff
@@ -0,0 +1,312 @@
+
+/// @file    AP_Mission_Relative.cpp
+/// @brief   translates and rotates Missions according to the location where Mode-AUTO has been switched on
+
+/*
+ *   The AP_Mission_Relative library
+ *   - memorizes the Location where Mode-AUTO has been switched on (Basepoint)
+ *   - moves the individual Waypoints according to Homepoint, Base-Point and Parameter-Settings
+ */
+
+#include <AP_Math/AP_Math.h>
+#include <AP_Common/Location.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <GCS_MAVLink/GCS.h>
+#include "AP_Mission_Relative.h"
+
+AP_Mission_Relative *AP_Mission_Relative::_singleton;
+
+const AP_Param::GroupInfo AP_Mission_Relative::var_info[] = {
+
+    // @Param: MOVE
+    // @DisplayName: kind of moving the Mission
+    // @Description: Defines how the Mission will be translated/rotated
+    // @Range: 0 5
+    // @Values: 0:No move, 1:PARALLEL translation, 2:additional ROTATION 1stWP-related, 3:additional ROTATION North-related, 4:additional ROTATION by Heading, 5:additional ROTATION by RC-channel
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("MOVE",  1, AP_Mission_Relative, _kind_of_move, AP_MISSION_RELATIVE_KIND_OF_MOVE_DEFAULT, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: RADIUS
+    // @DisplayName: radius to ignore translation
+    // @Description: Radius around HomeLocation therein translation of Mission will be ignored
+    // @Units: m
+    // @User: Advanced
+    AP_GROUPINFO("RADIUS",  2, AP_Mission_Relative, _no_translation_radius, AP_MISSION_RELATIVE_NO_TRANSLATION_RADIUS_DEFAULT),
+
+    // @Param: OPTIONS
+    // @DisplayName: altitude-offset options for translation
+    // @Description: Defines if and in which case altitude-offset will be used.
+    // @Values: 0:use Mission altitudes, 1:use altitude-offset Basepoint to first Waypoint for all Waypoints, 2:use altitude-offset only if positive
+    // @User: Advanced
+    AP_GROUPINFO("OPTIONS",  3, AP_Mission_Relative, _rel_options, AP_MISSION_RELATIVE_OPTIONS_DEFAULT),
+
+    // @Param: CHANNEL
+    // @DisplayName: rc-input-channel for rotation
+    // @Description: rc-input-channel to control amount of rotation 1000mys=-180deg 1500mys=0deg 2000mys=+180deg
+    // @Values: 0:no rc-input-channel selected
+    // @User: Advanced
+    AP_GROUPINFO("CHANNEL",  4, AP_Mission_Relative, _rotate_ch, 0),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_Mission_Relative::AP_Mission_Relative(void) {
+
+    AP_Param::setup_object_defaults(this, var_info); // load parameter defaults
+
+    _singleton = this;
+
+    _restart_behaviour = Restart_Behaviour::RESTART_NOT_TRANSLATED;
+    _translation.do_translation = false;
+    _translation.calculated = false;
+};
+
+void AP_Mission_Relative::memorize(ModeReason& mode_reason)
+{
+    if ((mode_reason != ModeReason::RC_COMMAND) && (mode_reason != ModeReason::GCS_COMMAND)) {
+        // just let the mission like it is (e.g. when soaring)
+        return;
+    }
+
+    switch (_kind_of_move) { // fix the behaviour at restart of the Mission
+    case 0:
```

Comment:
Instead of harcoding the case X values how about using the enum class Restart_Behaviour  here?

Reply:
I have to convert one time the "AP_Int8 Parameter _kind_of_move" into an "enum class Restart_Behaviour _restart_behaviour".
I do that in that switch, because e.g.
`_restart_behaviour = static_cast<Restart_Behaviour>(_kind_of_move);`
is not possible here.
But is there a more efficient methode to do that?

Reply:
hardcoded case-values replaced